# 01 - LSTM-GAN Training Notebook

This notebook trains the **baseline LSTM-GAN** model on the processed thruster 
time-series data generated in `00_data_preprocessing.ipynb`.

### **What this notebook accomplishes**
- Loads the processed NumPy datasets (`X_train_final.npy`, `X_val.npy`)
- Creates PyTorch Dataset & DataLoader objects
- Defines the LSTM-based Generator and Discriminator models
- Implements a stable GAN training loop (adversarial + reconstruction loss)
- Saves trained model checkpoints into the `models/` directory

### **Why this matters**
This is the foundation for:
- The upgraded **LSTM-VAE-GAN** architecture (Phase 4)
- Prophet residual fusion (Phase 5)
- Final anomaly detection (Notebook 02)

Once this notebook runs successfully, we will have a working GAN baseline.

In [3]:
# Imports & Device Setup

import os
import numpy as np
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)

Using device: cuda


## Load Processed Training & Validation Data

These arrays were saved in Notebook 00 and already contain:
- cleaned sequences
- normalized values
- sliding-window segments

We load only:
- `X_train_final.npy`
- `X_val.npy`

(`X_test.npy` will be generated later for Notebook 02)

In [4]:
# Load Data
DATA_DIR = "data/processed"

X_train = np.load(os.path.join(DATA_DIR, "X_train_final.npy"))
X_val   = np.load(os.path.join(DATA_DIR, "X_val.npy"))

print("X_train:", X_train.shape)
print("X_val:  ", X_val.shape)

seq_len   = X_train.shape[1]
n_features = X_train.shape[2]

seq_len, n_features

X_train: (304080, 128, 4)
X_val:   (76020, 128, 4)


(128, 4)

## PyTorch Dataset & DataLoader

We wrap the NumPy arrays into lightweight Dataset classes.
This makes the training loop efficient and batch-optimized.

We also set:
- `batch_size = 64` (safer for RAM/GPU)
- `shuffle=True` for training

In [5]:
# Dataset & DataLoader

class TimeSeriesDataset(Dataset):
    def __init__(self, X):
        self.X = torch.tensor(X, dtype=torch.float32)

    def __len__(self):
        return self.X.shape[0]

    def __getitem__(self, idx):
        return self.X[idx]

train_dataset = TimeSeriesDataset(X_train)
val_dataset   = TimeSeriesDataset(X_val)

train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True, drop_last=True)
val_loader   = DataLoader(val_dataset, batch_size=64, shuffle=False, drop_last=False)

len(train_dataset), len(val_dataset)

(304080, 76020)

## Building LSTM-GAN Components

We now define:

### **Generator**
- LSTM (input_dim → hidden_dim)
- Linear layer projecting back to input_dim
- Sigmoid activation since input is normalized

### **Discriminator**
- LSTM extracting temporal features
- Final FC layer → single real/fake probability
- Uses last hidden state for classification

In [6]:
# Model Definitions — Generator & Discriminator

class Generator(nn.Module):
    def __init__(self, input_dim, hidden_dim=128):
        super().__init__()
        self.lstm = nn.LSTM(input_dim, hidden_dim, batch_first=True)
        self.fc   = nn.Linear(hidden_dim, input_dim)

    def forward(self, z):
        out, _ = self.lstm(z)
        out = self.fc(out)
        return torch.sigmoid(out)


class Discriminator(nn.Module):
    def __init__(self, input_dim, hidden_dim=128):
        super().__init__()
        self.lstm = nn.LSTM(input_dim, hidden_dim, batch_first=True)
        self.fc   = nn.Linear(hidden_dim, 1)

    def forward(self, x):
        out, _ = self.lstm(x)
        last = out[:, -1, :]  # last timestep
        return torch.sigmoid(self.fc(last))


generator     = Generator(n_features, hidden_dim=128).to(device)
discriminator = Discriminator(n_features, hidden_dim=128).to(device)

generator, discriminator

(Generator(
   (lstm): LSTM(4, 128, batch_first=True)
   (fc): Linear(in_features=128, out_features=4, bias=True)
 ),
 Discriminator(
   (lstm): LSTM(4, 128, batch_first=True)
   (fc): Linear(in_features=128, out_features=1, bias=True)
 ))

## Loss Functions & Optimizers

We use:
- **BCE loss** for adversarial training
- **MSE loss** for reconstruction
- Adam optimizers for both networks

In [8]:
criterion_bce = nn.BCELoss()
criterion_mse = nn.MSELoss()

optimizer_G = torch.optim.Adam(generator.parameters(), lr=1e-4)
optimizer_D = torch.optim.Adam(discriminator.parameters(), lr=1e-4)

## Training Loop

The LSTM-GAN is trained for **30 epochs** in this baseline phase.

### Why 30 epochs?
- GANs (especially LSTM-based GANs) tend to **stabilize early**—most of the useful adversarial learning happens in the first 20–30 epochs.
- Beyond ~30 epochs, the Discriminator often becomes too strong, causing **generator collapse** or oscillations.
- This stage of the project only requires a **baseline GAN**, not a fully optimized model.  
  The stabilization and major performance improvements will come in the **LSTM-VAE-GAN phase**, where the encoder constrains the latent space and significantly improves training stability.
- Keeping this phase lightweight ensures fast iteration and lets us verify:
  - the architecture works  
  - losses behave properly  
  - gradients are stable  
  - the sequence shapes flow through correctly  

Once verified, the more advanced VAE-GAN (Phase 4) will be trained for longer (40–60 epochs) with far better stability.

---

### Training Routine (per batch)
1. **Train Discriminator**
   - Real batch → labeled as 1  
   - Fake batch (from Generator) → labeled as 0  
   - Update D to distinguish real vs. fake sequences.

2. **Train Generator**
   - Fool the discriminator (want D(fake) → 1)
   - Minimize reconstruction loss to keep generated sequences close to real ones

---

### What We Track During Training
- **Generator Loss** (adversarial + reconstruction)
- **Discriminator Loss** (real vs fake classification)
- **Validation Discriminator Score** (consistency check over time)

In [9]:
# Training Loop

num_epochs = 30

for epoch in range(1, num_epochs + 1):
    generator.train()
    discriminator.train()

    running_loss_D = 0.0
    running_loss_G = 0.0

    for real in train_loader:
        real = real.to(device)
        batch_size = real.size(0)

        
        # 1) Train Discriminator
        
        z = torch.randn(batch_size, seq_len, n_features, device=device)
        fake = generator(z).detach()

        real_labels = torch.ones(batch_size, 1, device=device)
        fake_labels = torch.zeros(batch_size, 1, device=device)

        d_real = discriminator(real)
        d_fake = discriminator(fake)

        loss_D = (criterion_bce(d_real, real_labels) +
                  criterion_bce(d_fake, fake_labels)) / 2

        optimizer_D.zero_grad()
        loss_D.backward()
        optimizer_D.step()

        
        # 2) Train Generator
        
        z = torch.randn(batch_size, seq_len, n_features, device=device)
        fake = generator(z)
        d_fake_for_G = discriminator(fake)

        loss_G_adv = criterion_bce(d_fake_for_G, real_labels)
        loss_G_rec = criterion_mse(fake, real)

        loss_G = loss_G_adv + loss_G_rec

        optimizer_G.zero_grad()
        loss_G.backward()
        optimizer_G.step()

        running_loss_D += loss_D.item() * batch_size
        running_loss_G += loss_G.item() * batch_size

    epoch_loss_D = running_loss_D / len(train_dataset)
    epoch_loss_G = running_loss_G / len(train_dataset)

    # quick validation sanity check
    generator.eval()
    discriminator.eval()
    with torch.no_grad():
        val_scores = []
        for x_val in val_loader:
            val_scores.append(discriminator(x_val.to(device)).mean().item())
        mean_val_score = np.mean(val_scores)

    print(f"Epoch {epoch:03d} | D_loss={epoch_loss_D:.4f} | "
          f"G_loss={epoch_loss_G:.4f} | D(val)={mean_val_score:.4f}")

Epoch 001 | D_loss=0.0488 | G_loss=6.7041 | D(val)=0.9975
Epoch 002 | D_loss=0.0028 | G_loss=8.3755 | D(val)=0.9995
Epoch 003 | D_loss=0.0047 | G_loss=9.9736 | D(val)=0.9995
Epoch 004 | D_loss=0.0058 | G_loss=10.6010 | D(val)=0.9989
Epoch 005 | D_loss=0.0120 | G_loss=10.7100 | D(val)=0.9988
Epoch 006 | D_loss=0.0002 | G_loss=11.9682 | D(val)=1.0000
Epoch 007 | D_loss=0.0271 | G_loss=12.4912 | D(val)=0.9926
Epoch 008 | D_loss=0.0010 | G_loss=11.7760 | D(val)=0.9997
Epoch 009 | D_loss=0.0001 | G_loss=12.9911 | D(val)=1.0000
Epoch 010 | D_loss=0.0000 | G_loss=13.8479 | D(val)=1.0000
Epoch 011 | D_loss=0.0000 | G_loss=14.2670 | D(val)=1.0000
Epoch 012 | D_loss=0.0000 | G_loss=14.3754 | D(val)=1.0000
Epoch 013 | D_loss=0.0000 | G_loss=16.1753 | D(val)=1.0000
Epoch 014 | D_loss=0.0000 | G_loss=16.6487 | D(val)=1.0000
Epoch 015 | D_loss=0.0000 | G_loss=18.0138 | D(val)=1.0000
Epoch 016 | D_loss=0.0000 | G_loss=19.1876 | D(val)=1.0000
Epoch 017 | D_loss=0.0000 | G_loss=20.7724 | D(val)=1.0000


## Why the Initial GAN Training Collapsed

The first training attempt produced the following pattern:

- **Discriminator loss fell to ~0.00 almost immediately**
- **Generator loss exploded** from ~6 → 24+
- **Validation discriminator score stayed ~0.99–1.00**
- Generator never improved across epochs

This behavior represents a classic failure mode in GAN training—**discriminator overpowering**.

### What Happened?

1. **The discriminator learned too quickly.**  
   Because real sequences have strong temporal structure and the generator initially 
   outputs uncorrelated noise, the discriminator can trivially classify real vs fake 
   from the very first batch.

2. **Discriminator saturated early.**  
   With D(real) → ~1.0 and D(fake) → ~0.0 consistently,  
   the discriminator's gradients became extremely small (near-zero BCE loss).

3. **Generator received no useful signal.**  
   Once D stops being “fooled,” the generator’s adversarial gradient vanishes.  
   As a result:
   - it cannot adjust toward the real distribution  
   - reconstruction loss dominates  
   - G_loss spikes dramatically  

4. **The GAN game became one-sided.**  
   GANs require a *balanced game* between G and D.  
   Here, D became perfect too early, so the generator was effectively training alone,
   leading to runaway loss.

### Why This Happens Specifically With LSTM-GANs

- LSTM-based discriminators are **very expressive**, even with small hidden sizes.  
- Time-series data has strong temporal continuity, which D learns faster than G can mimic.  
- Without regularization, D becomes overconfident and collapses the adversarial dynamics.

### Summary

The earlier training run failed not because the architecture was incorrect, but because the 
discriminator dominated the adversarial game too quickly. This produced:
- **D_loss ≈ 0.00**
- **G_loss → large and increasing**
- **No adversarial learning**
- **Complete GAN collapse**

This is a well-known issue in recurrent GANs and is resolved by weakening the discriminator,
applying label smoothing, and stabilizing gradients (as implemented in the corrected loop).

## Stabilized GAN Training — Preventing Discriminator Overpowering

During preliminary runs, the baseline LSTM-GAN showed classic *discriminator collapse*:
- Discriminator loss → extremely low (≈ 0.00)
- Generator loss → extremely high (6 → 20+)
- Validation score → always ~1.0
- Generator unable to improve or fool the discriminator

This behavior occurs when the discriminator becomes *too strong* in the early epochs,
leaving the generator with no meaningful gradient to learn from.

To stabilize training, we apply three widely-used and well-validated GAN techniques:

---

### **1. Label Smoothing (Soft Labels)**
Instead of labeling real as **1.0** and fake as **0.0**, we use:
- real → **0.9**
- fake → **0.1**

This prevents the discriminator from instantly saturating and forces it to retain uncertainty,
giving the generator space to improve.

---

### **2. Reduce Discriminator Capacity**
The discriminator is made intentionally weaker by using a smaller hidden dimension
(e.g., 64 instead of 128).  
This is standard GAN practice—if D is too expressive early on, GAN training becomes
one-sided and collapses.

---

### **3. Gradient Clipping for LSTM Stability**
Exploding gradients are common in recurrent GANs.
We apply:

nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)


for both G and D updates, ensuring stable, smooth learning dynamics.

---

### **Outcome**
These three modifications:
- prevent discriminator domination  
- stabilize generator updates  
- reduce GAN oscillation  
- allow meaningful adversarial learning even within 30 epochs  

This version forms a reliable baseline before extending the model to the **LSTM-VAE-GAN**
architecture in the next phase.

In [12]:
# FINAL — Generator & Discriminator (Stable Setup)

class Generator(nn.Module):
    def __init__(self, input_dim, hidden_dim=128):
        super().__init__()
        self.lstm = nn.LSTM(input_dim, hidden_dim, batch_first=True)
        self.fc   = nn.Linear(hidden_dim, input_dim)

    def forward(self, z):
        out, _ = self.lstm(z)
        out = self.fc(out)
        return torch.sigmoid(out)


class Discriminator(nn.Module):
    def __init__(self, input_dim, hidden_dim=64):   # intentionally weaker
        super().__init__()
        # removed dropout to avoid PyTorch warning
        self.lstm = nn.LSTM(input_dim, hidden_dim, batch_first=True)
        self.fc   = nn.Linear(hidden_dim, 1)

    def forward(self, x):
        out, _ = self.lstm(x)
        last = out[:, -1, :]
        return torch.sigmoid(self.fc(last))


generator     = Generator(n_features, hidden_dim=128).to(device)
discriminator = Discriminator(n_features, hidden_dim=64).to(device)

criterion_bce = nn.BCELoss()
criterion_mse = nn.MSELoss()

optimizer_G = torch.optim.Adam(generator.parameters(), lr=1e-4)
optimizer_D = torch.optim.Adam(discriminator.parameters(), lr=1e-4)

In [13]:
# Stabilized GAN Training Loop

num_epochs = 30

for epoch in range(1, num_epochs + 1):

    generator.train()
    discriminator.train()

    running_loss_D = 0.0
    running_loss_G = 0.0

    for real in train_loader:
        real = real.to(device)
        batch_size = real.size(0)

        
        # 1) Label Smoothing
        
        real_labels = torch.ones(batch_size, 1, device=device) * 0.9
        fake_labels = torch.zeros(batch_size, 1, device=device) + 0.1

        
        # 2) Train Discriminator
        
        z = torch.randn(batch_size, seq_len, n_features, device=device)
        fake = generator(z).detach()

        d_real = discriminator(real)
        d_fake = discriminator(fake)

        loss_D = (criterion_bce(d_real, real_labels) +
                  criterion_bce(d_fake, fake_labels)) / 2

        optimizer_D.zero_grad()
        loss_D.backward()

        # Prevent exploding gradients
        nn.utils.clip_grad_norm_(discriminator.parameters(), max_norm=1.0)

        optimizer_D.step()

        
        # 3) Train Generator
        
        z = torch.randn(batch_size, seq_len, n_features, device=device)
        fake = generator(z)

        d_fake_for_G = discriminator(fake)

        # adversarial + reconstruction
        loss_G_adv = criterion_bce(d_fake_for_G, real_labels)
        loss_G_rec = criterion_mse(fake, real)

        loss_G = loss_G_adv + loss_G_rec

        optimizer_G.zero_grad()
        loss_G.backward()

        # Clip generator gradients too
        nn.utils.clip_grad_norm_(generator.parameters(), max_norm=1.0)

        optimizer_G.step()

        running_loss_D += loss_D.item() * batch_size
        running_loss_G += loss_G.item() * batch_size

    
    # Epoch Summary
    
    epoch_loss_D = running_loss_D / len(train_dataset)
    epoch_loss_G = running_loss_G / len(train_dataset)

    generator.eval()
    discriminator.eval()
    with torch.no_grad():
        val_scores = []
        for x_val in val_loader:
            val_scores.append(discriminator(x_val.to(device)).mean().item())
        mean_val_score = np.mean(val_scores)

    print(f"Epoch {epoch:03d} | "
          f"D_loss={epoch_loss_D:.4f} | "
          f"G_loss={epoch_loss_G:.4f} | "
          f"D(val)={mean_val_score:.4f}")

Epoch 001 | D_loss=0.3409 | G_loss=3.1370 | D(val)=0.8995
Epoch 002 | D_loss=0.3268 | G_loss=3.2037 | D(val)=0.8608
Epoch 003 | D_loss=0.3280 | G_loss=3.2117 | D(val)=0.8998
Epoch 004 | D_loss=0.3251 | G_loss=3.2032 | D(val)=0.9000
Epoch 005 | D_loss=0.3251 | G_loss=3.1963 | D(val)=0.8999
Epoch 006 | D_loss=0.3251 | G_loss=3.1958 | D(val)=0.9000
Epoch 007 | D_loss=0.3260 | G_loss=3.1947 | D(val)=0.9006
Epoch 008 | D_loss=0.3252 | G_loss=3.1987 | D(val)=0.9000
Epoch 009 | D_loss=0.3251 | G_loss=3.1915 | D(val)=0.8999
Epoch 010 | D_loss=0.3251 | G_loss=3.1914 | D(val)=0.9000
Epoch 011 | D_loss=0.3251 | G_loss=3.1913 | D(val)=0.9001
Epoch 012 | D_loss=0.3257 | G_loss=3.1901 | D(val)=0.9005
Epoch 013 | D_loss=0.3251 | G_loss=3.1915 | D(val)=0.9000
Epoch 014 | D_loss=0.3251 | G_loss=3.1925 | D(val)=0.9001
Epoch 015 | D_loss=0.3251 | G_loss=3.1931 | D(val)=0.9002
Epoch 016 | D_loss=0.3251 | G_loss=3.1942 | D(val)=0.9000
Epoch 017 | D_loss=0.3251 | G_loss=3.1942 | D(val)=0.9000
Epoch 018 | D_

### Training Behaviour Summary (Baseline LSTM-GAN)

The stabilized training run shows the following behaviour over 30 epochs:

- **Discriminator loss** converges to ≈ **0.325**
- **Generator loss** converges to ≈ **3.19–3.21**
- **Validation discriminator output** (`D(val)`) remains around **0.90**

This indicates a healthy adversarial equilibrium:

- The discriminator is **better than random** (loss < 0.7) but not overpowering
  (loss ≈ 0, as in the collapsed run).
- The generator loss is **finite and stable**, with no explosive growth, which
  confirms that the generator receives meaningful gradients.
- The validation score staying near 0.90 suggests that the discriminator can
  still distinguish real-like structure in the sequences, without trivially
  rejecting all generated samples.

Compared to the earlier failed run (where `D_loss → 0.0`, `G_loss → 20+`,
and `D(val) → 1.0`), this training behaviour reflects a **balanced GAN game**.
The model has reached a reasonable Nash-type equilibrium, so 30 epochs are
sufficient for this baseline LSTM-GAN configuration. Further training would
likely yield diminishing returns and is not necessary for the baseline model.

In [14]:
# Save Trained Models

import os
os.makedirs("models", exist_ok=True)

gen_path = "models/generator_lstm_gan.pth"
disc_path = "models/discriminator_lstm_gan.pth"

torch.save(generator.state_dict(), gen_path)
torch.save(discriminator.state_dict(), disc_path)

print(f"Generator saved to: {gen_path}")
print(f"Discriminator saved to: {disc_path}")

Generator saved to: models/generator_lstm_gan.pth
Discriminator saved to: models/discriminator_lstm_gan.pth


## Baseline Anomaly Scoring (Reconstruction + Discriminator)

With the baseline LSTM-GAN trained and its weights saved, we now define a simple
unsupervised anomaly scoring scheme using two complementary signals:

1. **Reconstruction-style error (Generator-based)**  
   - For each input window `x`, we sample a noise tensor `z` of the same shape as `x`,  
     pass it through the trained Generator to obtain a fake sequence `x_fake`, and  
     compute a per-sample reconstruction-style error:
     
     \[
     \text{recon\_error}(x) = \frac{1}{T \cdot F} \sum_{t,f} (x_{t,f} - x^{\text{fake}}_{t,f})^2
     \]
     
     where \(T\) is the sequence length and \(F\) is the number of features.  
     Although the Generator is not a strict autoencoder, the imposed MSE term during
     training encourages it to produce sequences close to the normal data manifold.
     Large reconstruction error is therefore an indicator that the input window lies
     away from the learned “normal” region.

2. **Discriminator-based anomaly score**  
   - The Discriminator outputs a probability \(D(x) \in [0, 1]\) that an input
     sequence is “real” (i.e., consistent with the training distribution).
   - We convert this to an anomaly-like score by:
     
     \[
     \text{disc\_anom}(x) = 1 - D(x)
     \]
     
     so that higher values imply the discriminator is less confident that the window
     is normal.

3. **Hybrid anomaly score**  
   - A simple baseline combination is a weighted sum of both terms:
     
     \[
     \text{score}(x) = \alpha \cdot \text{recon\_error}(x)
                     + \beta  \cdot \text{disc\_anom}(x)
     \]
     
   - For this baseline, we start with \(\alpha = 1.0\) and \(\beta = 1.0\). These
     weights can later be tuned on a validation set or via downstream ROC/F1 analysis.

We compute and store these scores for:
- the **training set** (to understand the “normal” score distribution), and
- the **validation set** (to compare how scores shift on unseen-but-normal data).

These baseline scores will be used later in:
- the evaluation notebook (`02_evaluation.ipynb`),
- comparison against the LSTM-VAE-GAN variant,
- and for integrating with Prophet residuals in the final hybrid anomaly detector.

In [15]:
# Baseline Anomaly Scoring (Train + Val)

import numpy as np
import os

def compute_anomaly_scores(generator, discriminator, dataloader, device, alpha=1.0, beta=1.0):
    generator.eval()
    discriminator.eval()

    all_recon = []
    all_disc  = []
    all_hybrid = []

    with torch.no_grad():
        for real in dataloader:
            real = real.to(device)  # (B, T, F)
            batch_size = real.size(0)

            # Generator-based reconstruction-style error
            z = torch.randn(batch_size, seq_len, n_features, device=device)
            fake = generator(z)

            # Per-sample MSE over (T, F)
            recon_error = torch.mean((real - fake) ** 2, dim=(1, 2))  # (B,)

            # Discriminator-based anomaly term (1 - D(real))
            d_out = discriminator(real).squeeze(1)  # (B,)
            disc_anom = 1.0 - d_out                 # higher → more anomalous

            # Hybrid score
            hybrid = alpha * recon_error + beta * disc_anom

            all_recon.append(recon_error.cpu().numpy())
            all_disc.append(disc_anom.cpu().numpy())
            all_hybrid.append(hybrid.cpu().numpy())

    all_recon  = np.concatenate(all_recon, axis=0)
    all_disc   = np.concatenate(all_disc, axis=0)
    all_hybrid = np.concatenate(all_hybrid, axis=0)

    return {
        "recon_error": all_recon,
        "disc_anom": all_disc,
        "hybrid_score": all_hybrid,
    }


# Compute scores for train and val
train_scores = compute_anomaly_scores(generator, discriminator, train_loader, device)
val_scores   = compute_anomaly_scores(generator, discriminator, val_loader, device)

print("Train recon_error  mean/std:", train_scores["recon_error"].mean(), train_scores["recon_error"].std())
print("Train disc_anom    mean/std:", train_scores["disc_anom"].mean(), train_scores["disc_anom"].std())
print("Train hybrid_score mean/std:", train_scores["hybrid_score"].mean(), train_scores["hybrid_score"].std())
print("Val   recon_error  mean/std:", val_scores["recon_error"].mean(), val_scores["recon_error"].std())
print("Val   disc_anom    mean/std:", val_scores["disc_anom"].mean(), val_scores["disc_anom"].std())
print("Val   hybrid_score mean/std:", val_scores["hybrid_score"].mean(), val_scores["hybrid_score"].std())

# Optionally save to results/
os.makedirs("results", exist_ok=True)
np.save("results/train_recon_error.npy",  train_scores["recon_error"])
np.save("results/train_disc_anom.npy",    train_scores["disc_anom"])
np.save("results/train_hybrid_score.npy", train_scores["hybrid_score"])

np.save("results/val_recon_error.npy",    val_scores["recon_error"])
np.save("results/val_disc_anom.npy",      val_scores["disc_anom"])
np.save("results/val_hybrid_score.npy",   val_scores["hybrid_score"])

print("Saved train/val anomaly scores to results/")

Train recon_error  mean/std: 1.1180005 0.7218693
Train disc_anom    mean/std: 0.10000502 4.97567e-05
Train hybrid_score mean/std: 1.2180054 0.72184473
Val   recon_error  mean/std: 1.2840474 0.8195378
Val   disc_anom    mean/std: 0.1000118 5.9040547e-05
Val   hybrid_score mean/std: 1.3840593 0.8195067
Saved train/val anomaly scores to results/


### Baseline Anomaly Score Statistics

Using the trained baseline LSTM-GAN, we computed reconstruction-style errors,
discriminator-based anomaly terms, and their hybrid combination for both the
training and validation sets.

**Train set:**
- Recon error: mean ≈ 1.12, std ≈ 0.72  
- Disc anomaly: mean ≈ 0.10, std ≈ 0.00005  
- Hybrid score: mean ≈ 1.22, std ≈ 0.72  

**Validation set:**
- Recon error: mean ≈ 1.28, std ≈ 0.82  
- Disc anomaly: mean ≈ 0.10, std ≈ 0.00006  
- Hybrid score: mean ≈ 1.38, std ≈ 0.82  

Key observations:

- The validation set shows **slightly higher mean reconstruction and hybrid
  scores** than the training set, which is expected since the model was only
  optimized on training windows.
- The discriminator anomaly term is tightly concentrated around 0.10 on both
  train and validation data, reflecting that \(D(x) \approx 0.9\) for normal
  windows and that the discriminator is consistent across seen and unseen
  normal data.
- The hybrid score behaves as designed, essentially capturing reconstruction
  difficulty plus a small discriminator-based penalty, and will serve as a
  baseline distribution for “normal” behaviour.

These statistics provide a reference distribution that can later be used for:
- threshold selection (e.g., \(\mu + k\sigma\) rules),
- comparison with the LSTM-VAE-GAN variant,
- and integration with Prophet-based residuals in the final hybrid detector.

# LSTM-VAE-GAN EXTENSION

In this section, we extend the baseline LSTM-GAN to an **LSTM-VAE-GAN** variant.
The goal is to introduce a structured latent space via a variational encoder, which
can improve reconstruction quality and stability for anomaly detection.

### Architecture Overview

1. **Encoder (LSTM-based)**
   - Input: normal time-series window \( x \in \mathbb{R}^{T \times F} \)
   - An LSTM encodes the sequence into a hidden vector \( h \in \mathbb{R}^{H} \)
   - Two linear layers map \( h \) to:
     - \( \mu \in \mathbb{R}^{Z} \) (mean)
     - \( \log \sigma^2 \in \mathbb{R}^{Z} \) (log-variance)
   - Using the reparameterization trick, we sample:
     \[
     z = \mu + \sigma \odot \varepsilon,\quad \varepsilon \sim \mathcal{N}(0, I)
     \]

2. **VAE Generator (Decoder)**
   - Input: latent vector \( z \in \mathbb{R}^{Z} \)
   - We repeat \( z \) over the time dimension to obtain a sequence of shape
     \( (T, Z) \), then feed it to an LSTM-based decoder.
   - A linear layer projects decoder outputs to the original feature dimension \( F \),
     followed by a sigmoid to match the normalized input range.
   - Output: reconstructed sequence \( \hat{x} \in \mathbb{R}^{T \times F} \).

3. **Discriminator (same structure as baseline)**
   - Takes either \( x \) (real) or \( \hat{x} \) (reconstructed/generated),
     and outputs \( D(x) \in [0,1] \), the probability of being real.

### Loss Functions

We jointly train **Encoder + Generator** and the **Discriminator** with three terms:

1. **Reconstruction Loss (MSE)**
   \[
   \mathcal{L}_{\text{recon}} = \mathbb{E}\big[\|x - \hat{x}\|_2^2\big]
   \]

2. **KL Divergence Loss**
   Regularizes \( q(z|x) \) towards the standard normal prior:
   \[
   \mathcal{L}_{\text{KL}} = -\tfrac{1}{2} \sum_{i=1}^{Z}
   \big(1 + \log \sigma_i^2 - \mu_i^2 - \sigma_i^2 \big)
   \]

3. **Adversarial Loss**
   - Discriminator: binary cross-entropy on real vs. fake windows.
   - Encoder+Generator: tries to fool the discriminator:
     \[
     \mathcal{L}_{\text{adv}} = \text{BCE}(D(\hat{x}), \text{real\_label})
     \]

The total loss for Encoder+Generator is:
\[
\mathcal{L}_{E+G} =
\mathcal{L}_{\text{recon}} +
\lambda_{\text{KL}} \mathcal{L}_{\text{KL}} +
\lambda_{\text{adv}} \mathcal{L}_{\text{adv}}
\]

For this baseline VAE-GAN experiment, we start with:
- \( \lambda_{\text{KL}} = 0.01 \)
- \( \lambda_{\text{adv}} = 1.0 \)

The aim in this phase is to:
- Implement the Encoder and modified Generator
- Add KL + hybrid loss
- Train the LSTM-VAE-GAN variant
- Compare its stability and reconstruction behavior to the baseline LSTM-GAN.

In [16]:
# LSTM-VAE-GAN Components: Encoder + VAE Generator

latent_dim      = 32   # size of latent z
enc_hidden_dim  = 128  # encoder LSTM hidden size
dec_hidden_dim  = 128  # decoder LSTM hidden size (VAE generator)

class Encoder(nn.Module):
    def __init__(self, input_dim, hidden_dim, latent_dim):
        super().__init__()
        self.lstm = nn.LSTM(input_dim, hidden_dim, batch_first=True)
        self.fc_mu     = nn.Linear(hidden_dim, latent_dim)
        self.fc_logvar = nn.Linear(hidden_dim, latent_dim)

    def forward(self, x):
        # x: (B, T, F)
        out, (h_n, c_n) = self.lstm(x)      # h_n: (1, B, H)
        h = h_n.squeeze(0)                 # (B, H)
        mu     = self.fc_mu(h)             # (B, Z)
        logvar = self.fc_logvar(h)         # (B, Z)
        return mu, logvar


class VAEGenerator(nn.Module):
    def __init__(self, latent_dim, hidden_dim, output_dim, seq_len):
        super().__init__()
        self.seq_len = seq_len
        self.lstm = nn.LSTM(latent_dim, hidden_dim, batch_first=True)
        self.fc   = nn.Linear(hidden_dim, output_dim)

    def forward(self, z):
        # z: (B, Z)
        # repeat z across time dimension
        z_seq = z.unsqueeze(1).repeat(1, self.seq_len, 1)  # (B, T, Z)
        out, _ = self.lstm(z_seq)                         # (B, T, H)
        out = self.fc(out)                                # (B, T, F)
        return torch.sigmoid(out)


def reparameterize(mu, logvar):
    # mu, logvar: (B, Z)
    std = torch.exp(0.5 * logvar)
    eps = torch.randn_like(std)
    return mu + eps * std


# Instantiate VAE-GAN components
encoder_vae      = Encoder(n_features, enc_hidden_dim, latent_dim).to(device)
generator_vae    = VAEGenerator(latent_dim, dec_hidden_dim, n_features, seq_len).to(device)
discriminator_vae = Discriminator(n_features, hidden_dim=64).to(device)  # reuse same structure as before

# Loss + optimizers
criterion_bce = nn.BCELoss()
criterion_mse = nn.MSELoss()

lambda_kl   = 0.01
lambda_adv  = 1.0

optimizer_D_vae  = torch.optim.Adam(discriminator_vae.parameters(), lr=1e-4)
optimizer_EG_vae = torch.optim.Adam(
    list(encoder_vae.parameters()) + list(generator_vae.parameters()),
    lr=1e-4
)

In [17]:
# LSTM-VAE-GAN Training Loop

num_epochs_vae = 30

for epoch in range(1, num_epochs_vae + 1):

    encoder_vae.train()
    generator_vae.train()
    discriminator_vae.train()

    running_loss_D = 0.0
    running_loss_EG = 0.0
    running_recon = 0.0
    running_kl = 0.0
    running_adv = 0.0

    for real in train_loader:
        real = real.to(device)
        batch_size = real.size(0)

        
        # Label smoothing
        
        real_labels = torch.ones(batch_size, 1, device=device) * 0.9
        fake_labels = torch.zeros(batch_size, 1, device=device) + 0.1

        
        # 1) Train Discriminator
        
        # Encode → reparam → decode
        
        mu, logvar = encoder_vae(real)
        z = reparameterize(mu, logvar)
        fake = generator_vae(z).detach()   # detach for D step

        d_real = discriminator_vae(real)
        d_fake = discriminator_vae(fake)

        loss_D_real = criterion_bce(d_real, real_labels)
        loss_D_fake = criterion_bce(d_fake, fake_labels)
        loss_D = (loss_D_real + loss_D_fake) / 2

        optimizer_D_vae.zero_grad()
        loss_D.backward()
        nn.utils.clip_grad_norm_(discriminator_vae.parameters(), max_norm=1.0)
        optimizer_D_vae.step()

        
        # 2) Train Encoder + Generator (VAE + GAN)
        
        mu, logvar = encoder_vae(real)
        z = reparameterize(mu, logvar)
        fake = generator_vae(z)

        d_fake_for_EG = discriminator_vae(fake)

        # Reconstruction
        loss_recon = criterion_mse(fake, real)

        # KL divergence
        kl_term = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp(), dim=1)  # (B,)
        loss_kl = kl_term.mean()

        # Adversarial (want fake → real)
        loss_adv = criterion_bce(d_fake_for_EG, real_labels)

        # Total Encoder + Generator loss
        loss_EG = loss_recon + lambda_kl * loss_kl + lambda_adv * loss_adv

        optimizer_EG_vae.zero_grad()
        loss_EG.backward()
        nn.utils.clip_grad_norm_(encoder_vae.parameters(), max_norm=1.0)
        nn.utils.clip_grad_norm_(generator_vae.parameters(), max_norm=1.0)
        optimizer_EG_vae.step()

        running_loss_D  += loss_D.item() * batch_size
        running_loss_EG += loss_EG.item() * batch_size
        running_recon   += loss_recon.item() * batch_size
        running_kl      += loss_kl.item() * batch_size
        running_adv     += loss_adv.item() * batch_size

    # Epoch-level averages
    N = len(train_dataset)
    epoch_loss_D  = running_loss_D / N
    epoch_loss_EG = running_loss_EG / N
    epoch_recon   = running_recon / N
    epoch_kl      = running_kl / N
    epoch_adv     = running_adv / N

    # Quick val D check
    encoder_vae.eval()
    generator_vae.eval()
    discriminator_vae.eval()
    with torch.no_grad():
        val_scores = []
        for x_val in val_loader:
            x_val = x_val.to(device)
            d_val = discriminator_vae(x_val)
            val_scores.append(d_val.mean().item())
        mean_val_score = np.mean(val_scores)

    print(
        f"[VAE-GAN] Epoch {epoch:03d} | "
        f"D_loss={epoch_loss_D:.4f} | "
        f"EG_loss={epoch_loss_EG:.4f} | "
        f"Recon={epoch_recon:.4f} | KL={epoch_kl:.4f} | Adv={epoch_adv:.4f} | "
        f"D(val)={mean_val_score:.4f}"
    )

[VAE-GAN] Epoch 001 | D_loss=0.3492 | EG_loss=2.8051 | Recon=0.7825 | KL=2.4594 | Adv=1.9980 | D(val)=0.8991
[VAE-GAN] Epoch 002 | D_loss=0.3254 | EG_loss=2.7951 | Recon=0.6846 | KL=2.8482 | Adv=2.0820 | D(val)=0.9004
[VAE-GAN] Epoch 003 | D_loss=0.3251 | EG_loss=2.7566 | Recon=0.6501 | KL=2.3705 | Adv=2.0827 | D(val)=0.9002
[VAE-GAN] Epoch 004 | D_loss=0.3251 | EG_loss=2.7387 | Recon=0.6319 | KL=2.4059 | Adv=2.0828 | D(val)=0.9000
[VAE-GAN] Epoch 005 | D_loss=0.3251 | EG_loss=2.7266 | Recon=0.6187 | KL=2.5586 | Adv=2.0823 | D(val)=0.9001
[VAE-GAN] Epoch 006 | D_loss=0.3251 | EG_loss=2.7238 | Recon=0.6153 | KL=2.5823 | Adv=2.0828 | D(val)=0.9000
[VAE-GAN] Epoch 007 | D_loss=0.3251 | EG_loss=2.7224 | Recon=0.6127 | KL=2.6857 | Adv=2.0828 | D(val)=0.8999
[VAE-GAN] Epoch 008 | D_loss=0.3251 | EG_loss=2.7181 | Recon=0.6069 | KL=2.8389 | Adv=2.0828 | D(val)=0.8999
[VAE-GAN] Epoch 009 | D_loss=0.3251 | EG_loss=2.7150 | Recon=0.6034 | KL=2.8957 | Adv=2.0827 | D(val)=0.9001
[VAE-GAN] Epoch 010

In [18]:
# Save VAE-GAN Weights

os.makedirs("models", exist_ok=True)

enc_path  = "models/encoder_vae.pth"
gen2_path = "models/generator_vae.pth"
disc2_path = "models/discriminator_vae.pth"

torch.save(encoder_vae.state_dict(), enc_path)
torch.save(generator_vae.state_dict(), gen2_path)
torch.save(discriminator_vae.state_dict(), disc2_path)

print("Saved Encoder to:", enc_path)
print("Saved VAE-Generator to:", gen2_path)
print("Saved VAE-Discriminator to:", disc2_path)
print("All Phase 4 model weights saved successfully.")

Saved Encoder to: models/encoder_vae.pth
Saved VAE-Generator to: models/generator_vae.pth
Saved VAE-Discriminator to: models/discriminator_vae.pth
All Phase 4 model weights saved successfully.


## Phase 4 — LSTM-VAE-GAN: Stability and Reconstruction Quality

With the LSTM-VAE-GAN variant trained, we now evaluate:

1. **Stability during training**
   - Check that:
     - Discriminator loss remains in a reasonable band (no collapse to 0 or 0.7+).
     - Encoder+Generator (EG) loss does not explode.
     - Reconstruction loss decreases over epochs and then stabilizes.
     - Validation discriminator output `D(val)` remains around a consistent level,
       indicating no catastrophic overfitting or collapse.

2. **Reconstruction quality vs. baseline LSTM-GAN**
   - For each window \( x \), we compute:
     \[
     \text{recon\_error}(x) = \frac{1}{T \cdot F} \sum_{t,f} (x_{t,f} - \hat{x}_{t,f})^2
     \]
     where \( \hat{x} \) is the reconstruction produced by:
     - the **baseline LSTM-GAN generator** (using its training setup), and
     - the **LSTM-VAE-GAN (Encoder + VAE-Generator)** via the reparameterized latent
       \( z \sim q(z|x) \).

   - We then compare:
     - mean and standard deviation of reconstruction errors on:
       - the **training set**, and
       - the **validation set**,
     for both models.

The expectation is that the **LSTM-VAE-GAN** should:
- provide **lower reconstruction error** on normal data (tighter modeling of the
  normal manifold),
- while maintaining stable adversarial dynamics similar to the baseline LSTM-GAN.

The statistics below summarize this comparison.

In [21]:
# Phase 4 Evaluation — Reconstruction Quality Comparison

import numpy as np

def compute_vae_recon_errors(encoder, generator, dataloader, device):
    encoder.eval()
    generator.eval()

    all_recon = []

    with torch.no_grad():
        for real in dataloader:
            real = real.to(device)  # (B, T, F)

            # Encode → sample z → decode
            mu, logvar = encoder(real)
            z = reparameterize(mu, logvar)
            recon = generator(z)

            # Per-sample MSE over time & features
            recon_error = torch.mean((real - recon) ** 2, dim=(1, 2))  # (B,)
            all_recon.append(recon_error.cpu().numpy())

    all_recon = np.concatenate(all_recon, axis=0)
    return all_recon


# Baseline recon errors (already saved earlier)
baseline_train_recon = np.load("results/train_recon_error.npy")
baseline_val_recon   = np.load("results/val_recon_error.npy")

# VAE-GAN recon errors (train + val)
vae_train_recon = compute_vae_recon_errors(encoder_vae, generator_vae, train_loader, device)
vae_val_recon   = compute_vae_recon_errors(encoder_vae, generator_vae, val_loader, device)

# Compute mean/std for both models
def stats(name, arr):
    return f"{name}: mean = {arr.mean():.4f}, std = {arr.std():.4f}, min = {arr.min():.4f}, max = {arr.max():.4f}"

print("Baseline LSTM-GAN Reconstruction Errors:")
print(stats("Train", baseline_train_recon))
print(stats("Val  ", baseline_val_recon))

print("\nLSTM-VAE-GAN Reconstruction Errors:")
print(stats("Train", vae_train_recon))
print(stats("Val  ", vae_val_recon))

# Optionally save VAE recon stats for Notebook 02
np.save("results/vae_train_recon_error.npy", vae_train_recon)
np.save("results/vae_val_recon_error.npy",   vae_val_recon)

print("\nSaved VAE-GAN recon error arrays to results")

Baseline LSTM-GAN Reconstruction Errors:
Train: mean = 1.1180, std = 0.7219, min = 0.3965, max = 6.0212
Val  : mean = 1.2840, std = 0.8195, min = 0.3960, max = 5.5048

LSTM-VAE-GAN Reconstruction Errors:
Train: mean = 0.5664, std = 0.4260, min = 0.2187, max = 4.9492
Val  : mean = 0.6713, std = 0.5100, min = 0.2222, max = 4.0941

Saved VAE-GAN recon error arrays to results


# Conclusion

This notebook successfully completed the full model-training stages of the project,
covering both the **baseline LSTM-GAN** (Phase 3) and the extended **LSTM-VAE-GAN**
architecture (Phase 4). The overall goal of this notebook was to build the core
sequence-generation and anomaly-modeling engines that will later be evaluated,
compared, and combined with forecasting-based signals in Notebook 02.

---

## Baseline LSTM-GAN

The baseline GAN was implemented, stabilized, trained, and evaluated:

- Implemented PyTorch Dataset and DataLoader for sliding-window time-series
- Designed and trained an **LSTM-based Generator** and **LSTM-based Discriminator**
- Added adversarial (BCE) + reconstruction (MSE) losses
- Resolved mode-collapse issues using:
  - label smoothing,
  - reduced discriminator capacity,
  - gradient clipping
- Achieved stable equilibrium:
  - **D_loss ≈ 0.325**, **G_loss ≈ 3.19**, **D(val) ≈ 0.90**
- Saved the trained baseline model checkpoints
- Computed baseline reconstruction and hybrid anomaly scores
  - Train recon mean ≈ **1.1180**
  - Val recon mean ≈ **1.2840**

This baseline serves as the primary reference model for comparison.

---

## Phase 4 — LSTM-VAE-GAN Extension

The model was extended with a variational encoder to obtain a structured latent
space and improved reconstruction quality.

- Implemented an LSTM-based **Encoder** outputting \( \mu \) and \( \log \sigma^2 \)
- Added **reparameterization trick** to sample latent variables \( z \)
- Built the **VAE-Generator**, decoding repeated-latent sequences via an LSTM
- Added KL divergence, reconstruction loss, and adversarial components
- Trained Encoder + Generator jointly against the Discriminator
- Achieved stable training dynamics with:
  - **Recon ≈ 0.57**, **KL ≈ 3–4**, **EG_loss ≈ 2.7**, **D(val) ≈ 0.90**
- Saved VAE-GAN